src/helm/proxy/clients/huggingface_client.py

```diff
@@ -113,34 +124,45 @@ def serve_request(self, raw_request: Dict[str, Any]):
         return {"completions": completions, "input_length": len(encoded_input.input_ids[0])}
 
 
+_servers_lock: Lock = Lock()
+_servers: Dict[str, HuggingFaceServer] = {}
+
+
+def _get_singleton_server(model_config: HuggingFaceModelConfig) -> HuggingFaceServer:
+    global _servers_lock
+    global _servers
+    with _servers_lock:
+        if model_config.model_id not in _servers:
+            _servers[model_config.model_id] = HuggingFaceServer(model_config)
+    return _servers[model_config.model_id]
+
+
 class HuggingFaceClient(Client):
     def __init__(self, cache_config: CacheConfig):
         self.cache = Cache(cache_config)
         self.model_server_instances: Dict[str, HuggingFaceServer] = {}
 
     def get_model_server_instance(self, model) -> HuggingFaceServer:
-        if model not in self.model_server_instances:
-            model_config = get_huggingface_model_config(model)
-            if model_config:
-                self.model_server_instances[model] = HuggingFaceServer(model_config)
-            elif model == "EleutherAI/gpt-j-6B":
-                self.model_server_instances[model] = HuggingFaceServer(
-                    HuggingFaceModelConfig.from_string("EleutherAI/gpt-j-6B")
-                )
+        model_config = get_huggingface_model_config(model)
+        # Special-case some models in so that users don't have to enable them with --enable-huggingface-models
+        if not model_config:
+            # Other HuggingFace hub models that we'll look up for you even if you didn't enable them via the flag
+            if model == "EleutherAI/gpt-j-6B":
+                model_config = HuggingFaceHubModelConfig.from_string("EleutherAI/gpt-j-6B")
             elif model == "huggingface/gpt2":
-                self.model_server_instances[model] = HuggingFaceServer(HuggingFaceModelConfig.from_string("gpt2"))
+                model_config = HuggingFaceHubModelConfig.from_string("gpt2")
             elif model == "bigcode/santacoder":
                 self.model_server_instances[model] = HuggingFaceServer(
-                    HuggingFaceModelConfig.from_string("bigcode/santacoder")
+                    HuggingFaceHubModelConfig.from_string("bigcode/santacoder")
                 )
             elif model == "huggingface/starcoder":
                 self.model_server_instances[model] = HuggingFaceServer(
-                    HuggingFaceModelConfig.from_string("bigcode/starcoder")
+                    HuggingFaceHubModelConfig.from_string("bigcode/starcoder")
```

Comment:
I think this should just be `model_config = HuggingFaceHubModelConfig.from_string("bigcode/starcoder")`